vertical-pod-autoscaler/enhancements/3913-customized-recomender-vpa.md

```diff
@@ -0,0 +1,272 @@
+  - [Test Plan](#test-plan)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Alternatives](#alternatives)
+- [Out of Scope](#out-out-scope)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Today, the current VPA recommends CPU/Mem requests and limits based on one recommender, 
+which recommends the future requests/limits based on the historical usage observed in a 
+rolling time window. As there is no universal recommendation policy that applies to all 
+types of workload, this KEP suggests supporting multiple customized recommenders in VPA.  
+Thus, users can run different recommenders for different workloads, as they may exhibit 
+very distinct resource usage behaviors.
+
+## Motivation
+
+A VPA is used to resize containers in pods when the actual CPU/memory usage of a container 
+is significantly different from the resources requested. Resource usage-based recommendation 
+is the basic approach that resizes containers according to the actual usage observed and is 
+implemented in the default VPA recommender. Users can configure the time window and a certain 
+percentile of observed usage in the past as the prediction of future requests/limits for CPU/memory.
+
+However, as containers running different types of workloads may have different resource usage patterns, 
+there is no universal policy that applies to all. The existing VPA recommender may not accurately 
+predict future resource usage when containers exhibit certain resource usage behaviors, 
+such as trending, periodically changing, or occasional spikes, resulting in significant 
+over-provisioning and OOM kills for microservices. Learning different types of resource usage 
+behaviors for workloads and applying different algorithms to improve resource utilization 
+(CPU and Memory) predictions can significantly reduce over-provisioning and OOM kills in VPA.
+
+### Goals
+
+- Allow the VPA object to specify a customized recommender to use. 
+- Allow the VPA object to use the default recommender when no recommender is specified.
+
+### Non-Goals
+
+- We assume no pod uses two recommenders at the same time.
+- We do not resolve conflicts between recommenders.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 
+
+- Containers with Cyclic Patterns in Resource Usage
+
+Containers used in monitoring may receive load periodically to process but need to be long-running 
+to listen to incoming traffic. Thus, these containers usually exhibit cyclic patterns, alternating 
+between usage spikes and idling. Resizing containers according to usage observed in the previous 
+time window may always lead to under-provision for a short period when the load spikes just arrive. 
+The problem will even lead to frequent pod evictions when the pod was resized according to the idling 
+period and the host resource has been taken by other pods.
+
+#### Story 2 
+
+- Containers with Different but Recurrent Behaviors in Resource Usage
```

Comment:
I'm not sure I fully understand this one. If this is a single job, we don't have the ability to resize the workload between the phases. Currently default VPA will provision for the busiest phase, unless the phases are ~days long.